hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java

```diff
@@ -744,18 +744,21 @@ protected void logStaticSQL() {
 		}
 	}
 
-	private CollectionLoader standardCollectionLoader;
+	private volatile CollectionLoader standardCollectionLoader;
 
 	@Override
 	public void initialize(Object key, SharedSessionContractImplementor session) throws HibernateException {
 //		getAppropriateInitializer( key, session ).initialize( key, session );
 		determineLoaderToUse( key, session ).load( key, session );
 	}
 
+	// lazily initialize instance field via 'double-checked locking'
+	// see https://en.wikipedia.org/wiki/Double-checked_locking on why 'volatile' and local copy is used
 	protected CollectionLoader getStandardCollectionLoader() {
 		CollectionLoader localCopy = standardCollectionLoader;
 		if ( localCopy == null ) {
 			synchronized (this) {
+				localCopy = standardCollectionLoader;
```

Comment:
Without the above statement, there is no reason to check again at the next statement for `localCopy` would always be null.